### ISSUE TITLE
defaults channels leak into environments config

### ISSUE THREAD

**Issue by maximlt** [conda/conda#12136]
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

In this example the *defaults* channel is still available in the environment I create even if it has been removed from its configuration.

Executing `conda config --show-sources` from the *base* environment shows:

```
==> /Users/mliquet/.condarc <==
auto_activate_base: False
channel_priority: disabled
channels:
  - defaults
always_yes: False
experimental_solver: libmamba
```

Then I create an empty environment and change its config to add *conda-forge* and remove *defaults*:

```
conda create -n testleak
conda activate testleak
conda config --env --append channels conda-forge
conda config --env --remove channels defaults
```

Output of `conda config --show-sources` from the activated environment:

```
==> /Users/mliquet/.condarc <==
auto_activate_base: False
channel_priority: disabled
channels:
  - defaults
always_yes: False
experimental_solver: libmamba

==> /Users/mliquet/miniconda3/envs/testleak/.condarc <==
channels:
  - conda-forge
```

Now I try to install a package - `fonts-anaconda` - that is available on *defaults* channel but **not on *conda-forge***. I would expect the install to fail with a PackageNotFoundError, however it doesn't:

```
conda install fonts-anaconda

...
The following NEW packages will be INSTALLED:

  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0 None
  font-ttf-inconsol~ conda-forge/noarch::font-ttf-inconsolata-3.000-h77eed37_0 None
  font-ttf-source-c~ conda-forge/noarch::font-ttf-source-code-pro-2.038-h77eed37_0 None
  font-ttf-ubuntu    conda-forge/noarch::font-ttf-ubuntu-0.83-hab24e00_0 None
  fonts-anaconda     pkgs/main/noarch::fonts-anaconda-1-h8fa9717_0 None
...
```

### Conda Info

```shell
active environment : None
            shell level : 0
       user config file : /Users/mliquet/.condarc
 populated config files : /Users/mliquet/.condarc
          conda version : 22.9.0
    conda-build version : not installed
         python version : 3.9.5.final.0
       virtual packages : __osx=10.16=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /Users/mliquet/miniconda3  (writable)
      conda av data dir : /Users/mliquet/miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /Users/mliquet/miniconda3/pkgs
                          /Users/mliquet/.conda/pkgs
       envs directories : /Users/mliquet/miniconda3/envs
                          /Users/mliquet/.conda/envs
               platform : osx-64
             user-agent : conda/22.9.0 requests/2.27.1 CPython/3.9.5 Darwin/21.4.0 OSX/10.16 solver/libmamba conda-libmamba-solver/22.8.1 libmambapy/0.24.0
                UID:GID : 502:20
             netrc file : None
           offline mode : False
```

### Additional Context

The HoloViz group has a developer tool called `pyctdev` that is meant to be a unified interface for managing a Python project (e.g. `doit test_unit`, `doit build_package`, `doit build_docs`, ...). The user can either choose conda or pip as their dependency manager, i.e. `pyctdev` knows how to parse and convert the dependencies listed in a setup.py file so that they can be installed with conda. We're trying to see how we can improve this tool and having it declare the channels an environment should use is a pretty cool idea.

This issue might be related to https://github.com/conda/conda/issues/11874. If so, I'm happy to copy my OP there to add some more context.

**Comment by jaimergp** [comment]
I think this (arguably counterintuitive) behaviour is the expected one. See [this part of the docs](https://docs.conda.io/projects/conda/en/latest/user-guide/configuration/use-condarc.html#conflict-merging-strategy).

According to that, the two `channels` list are being _merged_; the env-specific config does _not_ override the user config... by default. I think you can make a config key "final" (e.g. no merging) by adding a special comment next to it: `#!final` ([see this blog post](https://www.anaconda.com/blog/conda-configuration-engine-power-users)).

This should be covered in the documentation, [but that's still pending](https://github.com/conda/conda/issues/11377).

**Comment by maximlt** [comment]
Thanks a lot @jaimergp for your answer. I did not know about the merging strategy, the precedence of the various config options and `#!final` ! Learnt a ton, that's great.

But now my next question is, does `conda config` allow to set `#!final` in an environment .condarc? I think what I'd want is (note the use of `#!final` in the local .condarc):
```
==> /Users/mliquet/.condarc <==
auto_activate_base: False
channel_priority: disabled
channels:
  - defaults
always_yes: False
experimental_solver: libmamba

==> /Users/mliquet/miniconda3/envs/testleak/.condarc <==
channels: #!final
  - conda-forge
```

**Comment by jaimergp** [comment]
Yes, any .condarc should support it. Whether it works or not it's up for debate because it's not documented (or even tested?) afaik. Try reproducing that issue and let's see what channels are loaded in operations in the new env.

**Comment by maximlt** [comment]
Sorry I wasn't very clear. What conda command should I execute to create this .condarc?
```
==> /Users/mliquet/miniconda3/envs/testleak/.condarc <==
channels: #!final
  - conda-forge
```

**Comment by jaimergp** [comment]
I see. So if I am understanding correctly, so `#!final` prevents a key being overridden by a higher precedence file, but it does not do the opposite (force-override an existing key)?

Makes sense from a sysadmin point of view (they would want to control what end users have in their config), but it's very inconvenient for single user setups like yours.

I'd label this as a feature request in that case, but we need to assess how to make everyone happy here; e.g. to prevent security risks in corporate / multi-user setups, a hypothetical `#!override` cannot be used if `#!final` is already set in that key up above in the condarc hierarchy.

Just a heads up this might take a while because there's plenty of backlog in the config system alone (see #12355 for an overview).

---

Side note: I also think `#!final` should be shown in the `conda config` output.

**Comment by maximlt** [comment]
> I see. So if I am understanding correctly, so #!final prevents a key being overridden by a higher precedence file, but it does not do the opposite (force-override an existing key)?

Oh yes indeed, I thought it was an override but it's instead telling the system not too look further down and stop building the config there, as in this example:

<img width="807" alt="image" src="https://user-images.githubusercontent.com/35924738/219341190-56dc4d95-9fa4-487b-87e7-59f17e9822e9.png">

And indeed if I add `#!final` to my default .condarc that just lists *defaults* then I can't install anything from any other channel.

As you say that all makes sense from the point of view of a sys admin!

> Just a heads up this might take a while because there's plenty of backlog in the config system alone (see https://github.com/conda/conda/issues/12355 for an overview).

Ush there's a lot in there! No worries this doesn't block us in any way.

**Comment by jaimergp** [comment]
Side note bis: since this will require some work, I guess we can get away with using the `#!` marks (making comments carry programmatic meaning is widespread in conda but that doesn't make it a good idea), and instead resort to some categorization:

```yaml
final:  # these keys won't be overriden
 channels:
   - conda-forge
```
```yaml
override:  # these keys won't be merged
 channels:
   - win-64
```

**Comment by jaimergp** [comment]
cc @travishathaway for visibility :)

**Comment by travishathaway** [comment]
Still not stale! 🍞

**Comment by travishathaway** [comment]
Still not stale! 🍞

Although, it's been so long since this was posted, I'm not sure whether we're actually going to change this behavior.
